src/main/java/com/google/sps/DataGraph.java

```diff
@@ -167,7 +201,7 @@ public boolean mutateGraph(Mutation mut) {
         }
         // Check whether any successor will have no in-edges after this node is removed
         // If so, make them roots
-        Set<GraphNode> successors = graph.successors(startNode);
+        successors = graph.successors(startNode);
```

Comment:
nit: you can just call graph.successors in the for loop header

Reply:
We were doing that originally. The issue is that before adding edges to the new node we have to delete the old node (and all edges to/from it) so when we get to the loop we can't access the successors and predecessors of the old node anymore.

Reply:
By loop header do you mean before the switch statement?

Reply:
I just tested it, and f i'm understanding this correctly, it's because the successors are only relevant in certain cases. So if the mutation is in adding a node and you call graph.successors(startNode) (and start not has yet to exist) you run into weird issues. (please correct me if my understanding of loop header is completely wrong!!)

Reply:
Oops @EatingW I misread the location of the loop, my response was in reference to another switch case. @haoyxc I think Yiting meant in the for-each loop like `for (GraphNode succ : graph.successors(startNode))`?